File: frontend/src/types/beneficiary.ts
export type BeneficiaryStatus = "ACTIVE" | "INACTIVE";

export interface Beneficiary {
  id: string;
  customerId: string;
  beneficiaryName: string;
  accountNumber: string;
  ifsc: string;
  bankName: string;
  nickname: string | null;
  status: BeneficiaryStatus;
  createdAt: string;
  updatedAt: string;
}

/** No `customerId` field — the owner is always the authenticated caller, derived from the JWT server-side. */
export interface CreateBeneficiaryRequest {
  beneficiaryName: string;
  accountNumber: string;
  ifsc: string;
  bankName: string;
  nickname?: string;
}

/**
 * Deliberately excludes `accountNumber`/`ifsc` — the backend rejects
 * changing them (see beneficiary-service's `UpdateBeneficiaryRequest`
 * javadoc: identifying fields aren't editable, only display fields are).
 */
export interface UpdateBeneficiaryRequest {
  beneficiaryName: string;
  bankName: string;
  nickname?: string;
}

File: backend/services/beneficiary-service/src/main/java/com/banksphere/beneficiary/entity/Beneficiary.java
package com.banksphere.beneficiary.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "beneficiaries")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Beneficiary {

    @Id
    @Builder.Default
    private UUID id = UUID.randomUUID();

    /** Owning customer, always taken from the caller's own JWT — never client-supplied. See docs/security/authorization.md. */
    @Column(name = "customer_id", nullable = false)
    private UUID customerId;

    @Column(name = "beneficiary_name", nullable = false, length = 100)
    private String beneficiaryName;

    @Column(name = "account_number", nullable = false, length = 20)
    private String accountNumber;

    @Column(name = "ifsc", nullable = false, length = 11)
    private String ifsc;

    @Column(name = "bank_name", nullable = false, length = 150)
    private String bankName;

    @Column(name = "nickname", length = 50)
    private String nickname;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    @Builder.Default
    private BeneficiaryStatus status = BeneficiaryStatus.ACTIVE;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        Instant now = Instant.now();
        this.createdAt = now;
        this.updatedAt = now;
        if (this.status == null) {
            this.status = BeneficiaryStatus.ACTIVE;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }
}
